### DISCUSSION TITLE
Displaying notifications on Android when the app is foregrounded

### DISCUSSION THREAD

**Discussion by munkii** [shinyorg/shiny#912] [Q&A] [answered]
I get that PushDelegate.OnEntry does not get called on Android. 

My understanding of "data" notifications and "Notification" notifications is that they are controlled by whether or not you are including the "notification" object in your JSON payload (I am sending these notifications via Azure Notification Hub). Ones without the "notifcation" object could be considered silent.

If the app is in the background and the notification is not a silent "data only" one then the Notification goes straight into the System tray.  If the app is in the foreground  the OS does not send it to the system tray instead it calls the app and in the case of Shiny PushDelegate OnReceived gets called.

What I really want is for a Notification Alert to popup when the app is in the foreground or the background. A couple of threads I have read say that I should be sending data only Push Notifications and then in the OnReceived callback creating local notifications. 

I'm happy with that as an approach but my attempts at creating local notifications still have them just appearing in the system tray!

I have tried the Shiny "Notifcation" Samples and they also just go straight to the System Tray

```
await shinyNotificationManager.Send(
    new Shiny.Notifications.Notification()
    {
        Message = notification.Notification.Message,
        Title = notification.Notification.Title,
        Android = new Shiny.Notifications.AndroidOptions() { LaunchActivityFlags = Shiny.Notifications.AndroidActivityFlags.SingleTop },
    });
```

[This SO question](https://stackoverflow.com/questions/68705044/android-app-not-showing-azure-notification-hub-notification-while-in-foreground) suggested that I need to create a Channel.

I quickly tried that but that still had them going into the System Tray

```
await shinyNotificationManager.AddChannel(
                                                        new Shiny.Notifications.Channel()
                                                        {
                                                            Importance = Shiny.Notifications.ChannelImportance.Normal,
                                                            Description = "General",
                                                            Identifier = "1",
                                                        });
```

Can someone please tell me where I am going wrong

**Comment by aritchie** [comment, marked as answer]
Notifications don't really happen if your app is in the foreground (especially on iOS).  From purely an android perspective, there may be some channel settings to allow this, but I'm not aware of what they are.  You are best to use the foreground events like pushManager.WhenReceived() or in your delegate, send a message for your UI to do something (toast, etc).  

As an FYI, if you do want to use the channel you've created, you also need to set it on your local notification.  In your example code, it isn't being set

**Reply by munkii** [reply]
> Notifications don't really happen if your app is in the foreground (especially on iOS)

That confuses me because as an iOS user I see notifcations for our app, when the app is in the foreground without actually doing anything special at all.

![IMG_1627](https://user-images.githubusercontent.com/827342/155306583-19a9228b-0334-4300-ae36-68b097d6a4ff.jpg)

> send a message for your UI to do something (toast, etc).
For the sake on consistency and reliability I think toast is the way we are going to go.

**Reply by munkii** [reply]
@aritchie I appreciate your probably a bit fed up of this topic but I can't thinbk of a better person to ask.

1. Can you help me with my confusion around the statement "Notifications don't really happen if your app is in the foreground (especially on iOS)". See my screenshot
2. If you want to reliable show a notifcation "popup" on Android and iOS would you send all notifcations silently and then simply to a Local Notification to alert the user if app backgrounded. If the app is Foregrounded resort to a custom Toast

Thanks

**Reply by aritchie** [reply]
1. Local untriggered (no schedule date, geofence, etc) notifications won't be displayed on iOS. 
2. If you only send silent notifications, iOS gives you roughly 5 an hour per device (they document 3-4, but it's slightly more).  If you want to display a toast in the foreground when a push is received, you can do something like the following:

```csharp
public class MyPushDelegate : IPushDelegate
{
    public MyPushDelegate(IPlatform platform, IYourToast toaster)  ...

    public async Task OnReceived(...) 
    {
        if (platform.Status == PlatformState.Foreground)
        {
             toaster.Show(...); // you will need to deal with issues like top window here
        }
    }
}
```

You can also use the foreground monitor for push in your viewmodel

```csharp
IPushManager manager; 

// make sure to dispose of the sub when you're done with it
// you'll also need to hop on the main thread
manager.WhenReceived().Subscribe(data => 
{
    // do a toast
});
```

**Reply by almirvuk** [reply]
Hi, @munkii, I am also using Shiny and Azure Notification Hub, have you managed to find a solution around this, to show the notification in both foreground and background mode on Android? :) Thanks!

**Reply by munkii** [reply]
@almirvuk a solution for Android? Yes.

**Background**
1. You must create a Channel. If you do not create a channel the app will use a the default one. The settings for the default app Channel are "Sound" whereas you want "Sound and pop-up". To get that I create a Channel in `PushDelegate.OnTokenRefreshed`

```
 await shinyNotificationManager.AddChannel(
                                                        new Shiny.Notifications.Channel()
                                                        {
                                                            Importance = Shiny.Notifications.ChannelImportance.High,
                                                            Description = "Default",
                                                            Identifier = "Default",
                                                        });

```
It is `Importance = Shiny.Notifications.ChannelImportance.High` that equates to "Sound and pop-up"


2. Next you need to target that Channel or at least that's what you'll want to do. However as you can see from this [StackOverflow question](https://stackoverflow.com/questions/71756681/cannot-target-android-channel-from-azure-notifcation-hub) I cannot get a notification to target a specfic Channel when using Azure Notifcation Hub.  I have an ongoing support ticket with Azure support about this. I thought I was going to be blocked until I saw a debug message saying 

> Missing Default Notification Channel metadata in AndroidManifest. Default value will be used.

From that I realised that if I just set the default channel in the AndroidManifest.xml my custom Channel would be used. I set the default like this

```
<application android:label="Breathe RM" android:icon="@mipmap/app_icon" Name="Breathe RM" tools:replace="android:label">
		<meta-data android:name="com.google.firebase.messaging.default_notification_channel_id" android:value="Default" />

```	

**Foreground**
1. The OnReceived callback  is fired on Android if the notifcation is not a silent one (i.e. is not using the data payload) and the App is in the foreground. From here we display a SnackBar at the top of the screen.

**Reply by almirvuk** [reply]
Nice, thank you! @munkii 

Do you use "content-available" : 1 -> in the iOS payload? :)

**Reply by munkii** [reply]
I do you use content-available : 1 in the payload when I am attempting to do background (silent) notifications. 
**As it currently stands I cannot get these to work in iOS at all when using "production" APNS environment and I do not understand why. "Sandbox" works fine** :-(

**Reply by almirvuk** [reply]
I am using the latest 2.5.x (preview) and facing the issue where OnEntry is not triggered on iOS, the Notification part of the PushNotification object struct is null for OnReceived.

**Reply by aritchie** [reply]
@almirvuk 

The Notification object will be null for background cases on both Android & iOS

      /// <summary>
      /// Notification properties received - this will be null if received in the background
      /// </summary>
      public Notification? Notification { get; }


@munkii 
Ensure both your server side & entitlements are set to production.  Also note that Apple throttles content available messages if you send too many (3-5) per hour on production, thus you will think things have stopped working if you've been testing too much

**Reply by almirvuk** [reply]
Hi, @aritchie thank you for the input, I am currently browsing the commits you had for 2.5.x and trying to understand why OnEntry is "ignored" on iOS in my case, "content-available": 1 in payload helps to trigger the OnReceived but after clicking the notification nothing happens both in foreground or background.

**Reply by aritchie** [reply]
@almirvuk I don't know.  I haven't fully tested everything on the iOS side yet.  I'd rather not discuss previews here as they simply aren't finished.  When I release to stable, if you are having an issue, please file it using the proper issue form: https://github.com/shinyorg/shiny/issues/new/choose   This forum is not meant for issues because there isn't enough info/code.

**Reply by almirvuk** [reply]
I agree 100%, you are right :) When it goes stable I will test it out and provide as much info as I can :) 

Thank you for all the great work!

**Comment by aritchie** [comment]
As of the next release 2.5 (currently in preview) - Foreground notifications will be handled by the library again.  You will only need to ensure the proper channel is passed (and created).

Also, onentry is now working fully by setting the click_action to SHINY_NOTIFICATION_CLICK and setting up an intent filter on your main activity

**Reply by almirvuk** [reply]
This is great! Thank you @aritchie :)

**Reply by almirvuk** [reply]
Wow! I tried the 2.5.0-preview-3108, and I can confirm that everything works as expected both foreground and background parts :) 

The only thing I noticed is that clicking on foreground notification re-creates a MainActivity, is that expected behavior or we should play with LaunchMode? :)

Thank you!
